File: src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.java
/*
 * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dv8tion.jda.api.hooks;

import net.dv8tion.jda.api.events.GenericEvent;
import net.dv8tion.jda.internal.JDAImpl;
import net.dv8tion.jda.internal.utils.ClassWalker;
import net.dv8tion.jda.internal.utils.JDALogger;
import org.slf4j.Logger;

import javax.annotation.Nonnull;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Implementation for {@link net.dv8tion.jda.api.hooks.IEventManager IEventManager}
 * which checks for {@link net.dv8tion.jda.api.hooks.SubscribeEvent SubscribeEvent} annotations on both
 * <b>static</b> and <b>member</b> methods.
 *
 * <p>Listeners for this manager do <u>not</u> need to implement {@link net.dv8tion.jda.api.hooks.EventListener EventListener}
 * <br>Example
 * <pre><code>
 * public class Foo
 * {
 *    {@literal @SubscribeEvent}
 *     public void onMsg(MessageReceivedEvent event)
 *     {
 *         System.out.printf("%s: %s\n", event.getAuthor().getName(), event.getMessage().getContentDisplay());
 *     }
 * }
 * </code></pre>
 *
 * @see net.dv8tion.jda.api.hooks.InterfacedEventManager
 * @see net.dv8tion.jda.api.hooks.IEventManager
 * @see net.dv8tion.jda.api.hooks.SubscribeEvent
 */
public class AnnotatedEventManager implements IEventManager
{
    private static final Logger LOGGER = JDALogger.getLog(AnnotatedEventManager.class);
    private final Set<Object> listeners = ConcurrentHashMap.newKeySet();
    private final Map<Class<?>, Map<Object, List<Method>>> methods = new ConcurrentHashMap<>();

    @Override
    public void register(@Nonnull Object listener)
    {
        if (listener.getClass().isArray())
        {
            for (Object o : ((Object[]) listener))
                register(o);
            return;
        }

        if (listeners.add(listener))
        {
            registerListenerMethods(listener);
        }
    }

    @Override
    public void unregister(@Nonnull Object listener)
    {
        if (listener.getClass().isArray())
        {
            for (Object o : ((Object[]) listener))
                unregister(o);
            return;
        }

        if (listeners.remove(listener))
        {
            updateMethods();
        }
    }

    @Nonnull
    @Override
    public List<Object> getRegisteredListeners()
    {
        return Collections.unmodifiableList(new ArrayList<>(listeners));
    }

    @Override
    public void handle(@Nonnull GenericEvent event)
    {
        for (Class<?> eventClass : ClassWalker.walk(event.getClass()))
        {
            Map<Object, List<Method>> listeners = methods.get(eventClass);
            if (listeners != null)
            {
                listeners.forEach((key, value) -> value.forEach(method ->
                {
                    try
                    {
                        method.setAccessible(true);
                        method.invoke(key, event);
                    }
                    catch (IllegalAccessException | InvocationTargetException e1)
                    {
                        JDAImpl.LOG.error("Couldn't access annotated EventListener method", e1);
                    }
                    catch (Throwable throwable)
                    {
                        JDAImpl.LOG.error("One of the EventListeners had an uncaught exception", throwable);
                        if (throwable instanceof Error)
                            throw (Error) throwable;
                    }
                }));
            }
        }
    }

    private void updateMethods()
    {
        methods.clear();
        for (Object listener : listeners)
        {
            registerListenerMethods(listener);
        }
    }

    private void registerListenerMethods(Object listener)
    {
        boolean isClass = listener instanceof Class;
        Class<?> c = isClass ? (Class<?>) listener : listener.getClass();
        Method[] allMethods = c.getDeclaredMethods();
        for (Method m : allMethods)
        {
            if (!m.isAnnotationPresent(SubscribeEvent.class))
                continue;
            //Skip member methods if listener is a Class
            if (isClass && !Modifier.isStatic(m.getModifiers()))
                continue;

            final Class<?>[] parameterTypes = m.getParameterTypes();
            if (parameterTypes.length != 1 || !GenericEvent.class.isAssignableFrom(parameterTypes[0]))
            {
                LOGGER.warn("Method '{}' annotated with @{} must have at most 1 parameter, which implements GenericEvent", m, SubscribeEvent.class.getSimpleName());
                continue;
            }

            Class<?> eventClass = parameterTypes[0];
            methods.computeIfAbsent(eventClass, k -> new ConcurrentHashMap<>())
                    .computeIfAbsent(listener, k -> new CopyOnWriteArrayList<>())
                    .add(m);
        }
    }
}

File: src/main/java/net/dv8tion/jda/api/hooks/AnnotatedEventManager.kt
/*
 * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dv8tion.jda.api.hooks

import net.dv8tion.jda.api.events.GenericEvent
import net.dv8tion.jda.internal.JDAImpl
import net.dv8tion.jda.internal.utils.ClassWalker
import net.dv8tion.jda.internal.utils.JDALogger
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer
import java.util.function.Function
import javax.annotation.Nonnull

/**
 * Implementation for [IEventManager][net.dv8tion.jda.api.hooks.IEventManager]
 * which checks for [SubscribeEvent][net.dv8tion.jda.api.hooks.SubscribeEvent] annotations on both
 * **static** and **member** methods.
 *
 *
 * Listeners for this manager do <u>not</u> need to implement [EventListener][net.dv8tion.jda.api.hooks.EventListener]
 * <br></br>Example
 * <pre>`
 * public class Foo
 * {
 * @SubscribeEvent
 * public void onMsg(MessageReceivedEvent event)
 * {
 * System.out.printf("%s: %s\n", event.getAuthor().getName(), event.getMessage().getContentDisplay());
 * }
 * }
`</pre> *
 *
 * @see net.dv8tion.jda.api.hooks.InterfacedEventManager
 *
 * @see net.dv8tion.jda.api.hooks.IEventManager
 *
 * @see net.dv8tion.jda.api.hooks.SubscribeEvent
 */
class AnnotatedEventManager : IEventManager {
    private val listeners: MutableSet<Any?> = ConcurrentHashMap.newKeySet()
    private val methods: MutableMap<Class<*>, MutableMap<Any?, MutableList<Method>>> = ConcurrentHashMap()
    override fun register(@Nonnull listener: Any) {
        if (listener.javaClass.isArray) {
            for (o in listener as Array<Any>) register(o)
            return
        }
        if (listeners.add(listener)) {
            registerListenerMethods(listener)
        }
    }

    override fun unregister(@Nonnull listener: Any?) {
        if (listener!!.javaClass.isArray) {
            for (o in (listener as Array<Any?>?)!!) unregister(o)
            return
        }
        if (listeners.remove(listener)) {
            updateMethods()
        }
    }

    @get:Nonnull
    override val registeredListeners: List<Any?>
        get() = Collections.unmodifiableList(ArrayList(listeners))

    override fun handle(@Nonnull event: GenericEvent) {
        for (eventClass in ClassWalker.walk(event.javaClass)) {
            val listeners: Map<Any?, MutableList<Method>>? = methods[eventClass]
            listeners?.forEach { (key: Any?, value: List<Method>) ->
                value.forEach(
                    Consumer { method: Method ->
                        try {
                            method.setAccessible(true)
                            method.invoke(key, event)
                        } catch (e1: IllegalAccessException) {
                            JDAImpl.LOG.error("Couldn't access annotated EventListener method", e1)
                        } catch (e1: InvocationTargetException) {
                            JDAImpl.LOG.error("Couldn't access annotated EventListener method", e1)
                        } catch (throwable: Throwable) {
                            JDAImpl.LOG.error("One of the EventListeners had an uncaught exception", throwable)
                            if (throwable is Error) throw throwable
                        }
                    })
            }
        }
    }

    private fun updateMethods() {
        methods.clear()
        for (listener in listeners) {
            registerListenerMethods(listener)
        }
    }

    private fun registerListenerMethods(listener: Any?) {
        val isClass = listener is Class<*>
        val c = if (isClass) listener as Class<*>? else listener!!.javaClass
        val allMethods = c!!.getDeclaredMethods()
        for (m in allMethods) {
            if (!m.isAnnotationPresent(SubscribeEvent::class.java)) continue
            //Skip member methods if listener is a Class
            if (isClass && !Modifier.isStatic(m.modifiers)) continue
            val parameterTypes = m.parameterTypes
            if (parameterTypes.size != 1 || !GenericEvent::class.java.isAssignableFrom(parameterTypes[0])) {
                LOGGER.warn(
                    "Method '{}' annotated with @{} must have at most 1 parameter, which implements GenericEvent",
                    m,
                    SubscribeEvent::class.java.getSimpleName()
                )
                continue
            }
            val eventClass = parameterTypes[0]
            methods.computeIfAbsent(
                eventClass,
                Function<Class<*>, MutableMap<Any?, MutableList<Method>>> { k: Class<*>? -> ConcurrentHashMap<Any?, List<Method>>() })
                .computeIfAbsent(listener) { k: Any? -> CopyOnWriteArrayList() }
                .add(m)
        }
    }

    companion object {
        private val LOGGER = JDALogger.getLog(AnnotatedEventManager::class.java)
    }
}
